[SPARK-4092] [CORE] Fix InputMetrics for coalesce'd Rdds

ksakellis opened the pull request:
When calculating the input metrics there was an assumption that one task only reads from one block - this is not true for some operations including coalesce. This patch simply increments the task's input metrics if previous ones existed of the same read method.

A limitation to this patch is that if a task reads from two different blocks of different read methods, one will override the other.

Commit:
[SPARK-4092] [CORE] Fix InputMetrics for coalesce'd Rdds

When calculating the input metrics there was an assumption
that one task only reads from one block - this is not true
for some operations including coalesce. This patch simply
increments the task's input metrics if previous ones existed
of the same read method.

A limitation to this patch is that if a task reads from
two different blocks of different read methods, one will override
the other.

Commit:
CR feedback

Commit:
Add bytesReadCallback to InputMetrics

Also added a test for interleaving reads.

pwendell commented:
Jenkins, test this please.

SparkQA commented:
[Test build #25404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25404/consoleFull) for   PR 3120 at commit [`a2ca793`](https://github.com/apache/spark/commit/a2ca793a97fb1fc0edafe417151bf674701fe7a4).
 * This patch merges cleanly.

pwendell commented on core/src/main/scala/org/apache/spark/CacheManager.scala:
```diff
@@ -44,7 +44,14 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
+        val prevBytesRead = existingMetrics
+          .filter(_.readMethod == blockResult.inputMetrics.readMethod)
+          .map(_.bytesRead)
+          .getOrElse(0L)
```
So what happens if we have input types that intermix here? For instance, what if they interleave between two input sources... will they just keep clobbering over eachother? It might be better to just chose a single input metric and stick with it, i.e. if we happen to be reading a block that wasn't derived from the same input as the one before it, just ignore it.

```
        val blockInput = blockResult.inputMetrics
        context.taskMetrics.inputMetrics match {
          case Some(existingInput) =>
            if (existingInput.readMethod == blockInput.readMethod) {
              existingInput.bytesRead += blockInput.bytesRead
            }
            // NOTE: If we have interleaving of two input types in one task, we currently ignore blocks associated
            //       with all but one type (whichever type was seen first). See SPARK-XXX.
          case None =>
            context.taskMetrics.inputMetrics = Some(blockInput)
        }
```

It's easier to document that behavior and also add a unit test for it.

pwendell replied:
Actually after looking at Hadoop RDD - it might be necessary to just clobber here to preserve consistency with that case. But it could still be nicer to write this with a match.

kayousterhout replied:
What if there are 3 input sources that interleave here?  Suppose you have (1) input from cache, (2) input from Hadoop, and (3) input from cache.  My understanding is that when (2) starts being read, it will clobber the input metrics from (1).  Then, when (3) is read, it will again clobber the input metrics, so the metrics won't properly reflect the total data read from cache (they'll only reflect the data read from (3)).  Is that right?

ksakellis replied:
Yes, when the cache is being used then yes there will be clobbering. There are a few solutions
1) we can not filter on readMethod and just append blindly. That way we don't override any metrics but the eventual read method will not be correct (either first wins or last wins - whatever we choose)
2) we model input metrics like we do with shuffle metrics where we collect an array of them and then finally we sum them up. - this is a bigger change.

kayousterhout replied:
Is it possible to do @pwendell 's suggestion, where you check the type of the input metrics and only append if it's the same type?  I'd actually be slightly in favor of just returning a list of input metrics, one for each input type, because the other solutions seem a little hacky -- but defer to @sryza / @pwendell here (who I think had argued in the past that this extra complexity wasn't worth it).

SparkQA commented:
[Test build #25404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25404/consoleFull) for   PR 3120 at commit [`a2ca793`](https://github.com/apache/spark/commit/a2ca793a97fb1fc0edafe417151bf674701fe7a4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25404/
Test FAILed.

pwendell commented on core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala:
```diff
@@ -153,34 +157,19 @@ class NewHadoopRDD[K, V](
         havePair = false
-
-        // Update bytes read metric every few records
-        if (recordsSinceMetricsUpdate == HadoopRDD.RECORDS_BETWEEN_BYTES_READ_METRIC_UPDATES
```
This was done intentionally to help keep the callback updates out of the `InputMetrics` class and isolate it to Hadoop RDD. This notion of callbacks makes the InputMetrics class more complicated and mutable. Since it's an exposed class we really wanted to keep the interface clean and simple, even if it meant some extra engineering in HadoopRDD.  So could this part of the change be reverted back to how it was before (and you don't change the InputMetrics/TaskMetrics classes?).

ksakellis replied:
@pwendell There is a long thread in this pr between @sryza and @kayousterhout about why we need to add the call back to the input metrics. The reason is to prevent clobbering between different HadoopRdds. For example CartesianRdd - this is why there is a specific unit test for that case. I don't think we can do anything correctly if we don't have the callbacks in the inputMetrics.

pwendell replied:
Okay, that's fine then. I looked and it's all private[spark] so actually there is no change to visibility.

pwendell commented on core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala:
```diff
@@ -109,18 +109,22 @@ class NewHadoopRDD[K, V](
       val conf = confBroadcast.value.value
 
-      val inputMetrics = new InputMetrics(DataReadMethod.Hadoop)
+      val readMethod = DataReadMethod.Hadoop
```
Minor, but I find it slightly easier to follow as a match:

```
      val hadoopReadMethod = DataReadMethod.Hadoop
      val newMetrics = context.taskMetrics.inputMetrics match {
        case Some(InputMetrics(hadoopReadMethod)) =>
          context.taskMetrics.inputMetrics
        case _ =>
          // Note that this may clobber some other input metric (see SPARK-XXX)
          hadoopReadMethod
      }
      context.taskMetrics.inputMetrics = Some(newMetrics)
```

kayousterhout commented on core/src/test/scala/org/apache/spark/metrics/InputOutputMetricsSuite.scala:
```diff
@@ -17,66 +17,147 @@
-
+import org.apache.spark.scheduler.{SparkListener, SparkListenerTaskEnd}
+import org.apache.spark.util.Utils
 import org.scalatest.FunSuite
```
Fix import ordering here (there should be a line break between each different org, and scalatest should come before Spark)

Commit:
Drops metrics if conflicting read methods exist

Tasks now only store/accumulate input metrics from
the same read method. If a task has interleaved reads
from more than one block of different read methods, we
choose to store the first read methods metrics.

https://issues.apache.org/jira/browse/SPARK-5225
addresses keeping track of all input metrics.

This change also centralizes this logic in TaskMetrics
and gates how inputMetrics can be added to TaskMetrics.

pwendell commented:
Jenkins, test this please.

pwendell commented:
LGTM pending tests

SparkQA commented:
[Test build #25623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25623/consoleFull) for   PR 3120 at commit [`54e6658`](https://github.com/apache/spark/commit/54e6658d937f2da1c7ab9a2f8412e3a0cfc420b3).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25623/consoleFull) for   PR 3120 at commit [`54e6658`](https://github.com/apache/spark/commit/54e6658d937f2da1c7ab9a2f8412e3a0cfc420b3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25623/
Test PASSed.

rxin commented on core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala:
```diff
@@ -146,6 +185,10 @@ class TaskMetrics extends Serializable {
     _shuffleReadMetrics = Some(merged)
   }
+
+  private[spark] def updateInputMetrics() = synchronized {
```
in your next pr, can u fix this by adding a return type explicitly?

ksakellis replied:
So this follows the method above: updateShuffleReadMetrics that doesn't have a return type. Should I change both then?

rxin replied:
would be great to do that!
